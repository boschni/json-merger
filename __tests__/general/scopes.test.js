const jsonMerger = require("../../dist");
const { testConfig } = require("../../__helpers__");

jest.mock("fs");
const fs = require("fs");

describe("when merging object1 and object2", function () {
  test("then $target in object1 should refer to undefined", function () {
    const object1 = {
      a: {
        $expression: "$target",
      },
    };

    const object2 = {};

    const result = jsonMerger.mergeObjects([object1, object2], testConfig());

    expect(result).toMatchSnapshot();
  });

  test("then $target in object2 should refer to the processed object1", function () {
    const object1 = {
      targetProp: "should be the value of sharedProp",
      sharedProp: "object1.sharedProp",
    };

    const object2 = {
      sharedProp: {
        $expression: "$target.targetProp",
      },
    };

    const result = jsonMerger.mergeObjects([object1, object2], testConfig());

    expect(result).toMatchSnapshot();
  });

  test("then $targetProperty in object1 should refer to undefined", function () {
    const object1 = {
      a: {
        $expression: "$targetProperty",
      },
    };

    const object2 = {};

    const result = jsonMerger.mergeObjects([object1, object2], testConfig());

    expect(result).toMatchSnapshot();
  });

  test("then $targetProperty in object2 should refer to the processed property in object1", function () {
    const object1 = {
      targetProp: "object1.targetProp",
      sharedProp: "should be the value of sharedProp",
    };

    const object2 = {
      sourceProp: "object2.sourceProp",
      sharedProp: {
        $expression: "$targetProperty",
      },
    };

    const result = jsonMerger.mergeObjects([object1, object2], testConfig());

    expect(result).toMatchSnapshot();
  });

  test("then $source in object2 should refer to the unprocessed object2", function () {
    const object1 = {
      targetProp: "object1.targetProp",
      sharedProp: "object1.sharedProp",
    };

    const object2 = {
      sourceProp: "should be the value of sharedProp",
      sharedProp: {
        $expression: "$source.sourceProp",
      },
    };

    const result = jsonMerger.mergeObjects([object1, object2], testConfig());

    expect(result).toMatchSnapshot();
  });

  test("then $sourceProperty in object2 should refer to the unprocessed object2 property", function () {
    const object1 = {
      targetProp: "object1.targetProp",
      sharedProp: "object1.sharedProp",
    };

    const object2 = {
      sharedProp: {
        $expression: "$sourceProperty",
      },
    };

    const result = jsonMerger.mergeObjects([object1, object2], testConfig());

    expect(result).toMatchSnapshot();
  });

  describe("and two objects are merged with $merge in object2", function () {
    test("then $root.$target in object2 $merge.with should refer to the processed object1", function () {
      const object1 = {
        targetProp: "should be the value of mergeSharedProp",
        sharedProp: "object1.sharedProp",
      };

      const object2 = {
        sourceProp: "object2.sourceProperty",
        sharedProp: {
          $merge: {
            source: {
              mergeSourceProp: "$merge.source.mergeSourceProp",
              mergeSharedProp: "$merge.source.mergeSharedProp",
            },
            with: {
              mergeWithProp: "$merge.with.mergeWithProp",
              mergeSharedProp: {
                $expression: "$root.$target.targetProp",
              },
            },
          },
        },
      };

      const result = jsonMerger.mergeObjects([object1, object2], testConfig());

      expect(result).toMatchSnapshot();
    });

    test("then $target in object2 $merge.with should refer to the processed object2 $merge.source", function () {
      const object1 = {
        targetProp: "object1.targetProp",
        sharedProp: "object1.sharedProp",
      };

      const object2 = {
        sourceProp: "object2.sourceProperty",
        sharedProp: {
          $merge: {
            source: {
              mergeSourceProp: "should be the value of mergeSharedProp",
              mergeSharedProp: "$merge.source.mergeSharedProp",
            },
            with: {
              mergeWithProp: "$merge.with.mergeWithProp",
              mergeSharedProp: {
                $expression: "$target.mergeSourceProp",
              },
            },
          },
        },
      };

      const result = jsonMerger.mergeObjects([object1, object2], testConfig());

      expect(result).toMatchSnapshot();
    });

    test("then $parent.$target in object2 $merge.with should refer to the processed object1", function () {
      const object1 = {
        targetProp: "should be the value of mergeSharedProp",
        sharedProp: "object1.sharedProp",
      };

      const object2 = {
        sourceProp: "object2.sourceProperty",
        sharedProp: {
          $merge: {
            source: {
              mergeSourceProp: "$merge.source.mergeSourceProp",
              mergeSharedProp: "$merge.source.mergeSharedProp",
            },
            with: {
              mergeWithProp: "$merge.with.mergeWithProp",
              mergeSharedProp: {
                $expression: "$root.$target.targetProp",
              },
            },
          },
        },
      };

      const result = jsonMerger.mergeObjects([object1, object2], testConfig());

      expect(result).toMatchSnapshot();
    });

    test("then $targetProperty in object2 $merge.with should refer to the processed object2 $merge.source property", function () {
      const object1 = {
        targetProp: "object1.targetProp",
        sharedProp: "object1.sharedProp",
      };

      const object2 = {
        sourceProp: "object2.sourceProperty",
        sharedProp: {
          $merge: {
            source: {
              mergeSourceProp: "$merge.source.mergeSourceProp",
              mergeSharedProp: "should be the value of mergeSharedProp",
            },
            with: {
              mergeWithProp: "$merge.with.mergeWithProp",
              mergeSharedProp: {
                $expression: "$targetProperty",
              },
            },
          },
        },
      };

      const result = jsonMerger.mergeObjects([object1, object2], testConfig());

      expect(result).toMatchSnapshot();
    });

    test("then $root.$source in object2 $merge.with should refer to the unprocessed object2", function () {
      const object1 = {
        targetProp: "object1.targetProp",
        sharedProp: "object1.sharedProp",
      };

      const object2 = {
        sourceProp: "should be the value of mergeSharedProp",
        sharedProp: {
          $merge: {
            source: {
              mergeSourceProp: "$merge.source.mergeSourceProp",
              mergeSharedProp: "$merge.source.mergeSharedProp",
            },
            with: {
              mergeWithProp: "$merge.with.mergeWithProp",
              mergeSharedProp: {
                $expression: "$root.$source.sourceProp",
              },
            },
          },
        },
      };

      const result = jsonMerger.mergeObjects([object1, object2], testConfig());

      expect(result).toMatchSnapshot();
    });

    test("then $source in object2 $merge.with should refer to the unprocessed object2 $merge.with", function () {
      const object1 = {
        targetProp: "object1.targetProp",
        sharedProp: "object1.sharedProp",
      };

      const object2 = {
        sourceProp: "object2.sourceProperty",
        sharedProp: {
          $merge: {
            source: {
              mergeSourceProp: "$merge.source.mergeSourceProp",
              mergeSharedProp: "$merge.source.mergeSharedProp",
            },
            with: {
              mergeWithProp: "should be the value of mergeSharedProp",
              mergeSharedProp: {
                $expression: "$source.mergeWithProp",
              },
            },
          },
        },
      };

      const result = jsonMerger.mergeObjects([object1, object2], testConfig());

      expect(result).toMatchSnapshot();
    });

    test("then $sourceProperty in object2 $merge.with should refer to the unprocessed object2 $merge.with property", function () {
      const object1 = {
        targetProp: "object1.targetProp",
        sharedProp: "object1.sharedProp",
      };

      const object2 = {
        sourceProp: "object2.sourceProperty",
        sharedProp: {
          $merge: {
            source: {
              mergeSourceProp: "$merge.source.mergeSourceProp",
              mergeSharedProp: "$merge.source.mergeSharedProp",
            },
            with: {
              mergeWithProp: "$merge.with.mergeWithProp",
              mergeSharedProp: {
                $expression: "$sourceProperty",
              },
            },
          },
        },
      };

      const result = jsonMerger.mergeObjects([object1, object2], testConfig());

      expect(result).toMatchSnapshot();
    });
  });
});

describe("when importing a file", function () {
  test("then $root in the file should refer to the root of the file", function () {
    const files = {
      "a.json": {
        aa: {
          $expression: "$root.$source.sharedProp",
        },
        sharedProp: "This should be the value of /a/aa",
      },
    };

    const object = {
      a: {
        $import: "a.json",
      },
      sharedProp: "wrong",
    };

    fs.__setJsonMockFiles(files);

    const result = jsonMerger.mergeObject(object, testConfig());

    expect(result).toMatchSnapshot();
  });

  test("then the parent scope variables should not be accessible", function () {
    const files = {
      "b.json": {
        $expression: "typeof $repeat",
      },
    };

    const object = {
      $repeat: {
        range: "1",
        value: {
          a: {
            $expression: "typeof $repeat",
          },
          b: {
            $import: "b.json",
          },
        },
      },
    };

    fs.__setJsonMockFiles(files);

    const result = jsonMerger.mergeObject(object, testConfig());

    expect(result).toMatchSnapshot();
  });

  test("then $sourceFilename in the file should refer to the filename without extension", function () {
    const files = {
      "a.json": {
        a: {
          $expression: "$sourceFileName",
        },
        b: {
          $import: "b.json",
        },
      },
      "b.json": {
        $expression: "$sourceFileName",
      },
    };

    fs.__setJsonMockFiles(files);

    const result = jsonMerger.mergeFile("a.json", testConfig());

    expect(result).toMatchSnapshot();
  });
});
