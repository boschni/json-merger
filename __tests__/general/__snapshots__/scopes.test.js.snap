// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when importing a file then $root in the file should refer to the root of the file 1`] = `
"{
	\\"a\\": {
		\\"aa\\": \\"This should be the value of /a/aa\\",
		\\"sharedProp\\": \\"This should be the value of /a/aa\\"
	},
	\\"sharedProp\\": \\"wrong\\"
}"
`;

exports[`when importing a file then $sourceFilename in the file should refer to the filename without extension 1`] = `
"{
	\\"a\\": \\"a\\",
	\\"b\\": \\"b\\"
}"
`;

exports[`when importing a file then the parent scope variables should not be accessible 1`] = `
"[
	{
		\\"a\\": \\"object\\",
		\\"b\\": \\"undefined\\"
	}
]"
`;

exports[`when merging object1 and object2 and two objects are merged with $merge in object2 then $parent.$target in object2 $merge.with should refer to the processed object1 1`] = `
"{
	\\"targetProp\\": \\"should be the value of mergeSharedProp\\",
	\\"sharedProp\\": {
		\\"mergeSourceProp\\": \\"$merge.source.mergeSourceProp\\",
		\\"mergeSharedProp\\": \\"should be the value of mergeSharedProp\\",
		\\"mergeWithProp\\": \\"$merge.with.mergeWithProp\\"
	},
	\\"sourceProp\\": \\"object2.sourceProperty\\"
}"
`;

exports[`when merging object1 and object2 and two objects are merged with $merge in object2 then $root.$source in object2 $merge.with should refer to the unprocessed object2 1`] = `
"{
	\\"targetProp\\": \\"object1.targetProp\\",
	\\"sharedProp\\": {
		\\"mergeSourceProp\\": \\"$merge.source.mergeSourceProp\\",
		\\"mergeSharedProp\\": \\"should be the value of mergeSharedProp\\",
		\\"mergeWithProp\\": \\"$merge.with.mergeWithProp\\"
	},
	\\"sourceProp\\": \\"should be the value of mergeSharedProp\\"
}"
`;

exports[`when merging object1 and object2 and two objects are merged with $merge in object2 then $root.$target in object2 $merge.with should refer to the processed object1 1`] = `
"{
	\\"targetProp\\": \\"should be the value of mergeSharedProp\\",
	\\"sharedProp\\": {
		\\"mergeSourceProp\\": \\"$merge.source.mergeSourceProp\\",
		\\"mergeSharedProp\\": \\"should be the value of mergeSharedProp\\",
		\\"mergeWithProp\\": \\"$merge.with.mergeWithProp\\"
	},
	\\"sourceProp\\": \\"object2.sourceProperty\\"
}"
`;

exports[`when merging object1 and object2 and two objects are merged with $merge in object2 then $source in object2 $merge.with should refer to the unprocessed object2 $merge.with 1`] = `
"{
	\\"targetProp\\": \\"object1.targetProp\\",
	\\"sharedProp\\": {
		\\"mergeSourceProp\\": \\"$merge.source.mergeSourceProp\\",
		\\"mergeSharedProp\\": \\"should be the value of mergeSharedProp\\",
		\\"mergeWithProp\\": \\"should be the value of mergeSharedProp\\"
	},
	\\"sourceProp\\": \\"object2.sourceProperty\\"
}"
`;

exports[`when merging object1 and object2 and two objects are merged with $merge in object2 then $sourceProperty in object2 $merge.with should refer to the unprocessed object2 $merge.with property 1`] = `
"{
	\\"targetProp\\": \\"object1.targetProp\\",
	\\"sharedProp\\": {
		\\"mergeSourceProp\\": \\"$merge.source.mergeSourceProp\\",
		\\"mergeSharedProp\\": \\"$sourceProperty\\",
		\\"mergeWithProp\\": \\"$merge.with.mergeWithProp\\"
	},
	\\"sourceProp\\": \\"object2.sourceProperty\\"
}"
`;

exports[`when merging object1 and object2 and two objects are merged with $merge in object2 then $target in object2 $merge.with should refer to the processed object2 $merge.source 1`] = `
"{
	\\"targetProp\\": \\"object1.targetProp\\",
	\\"sharedProp\\": {
		\\"mergeSourceProp\\": \\"should be the value of mergeSharedProp\\",
		\\"mergeSharedProp\\": \\"should be the value of mergeSharedProp\\",
		\\"mergeWithProp\\": \\"$merge.with.mergeWithProp\\"
	},
	\\"sourceProp\\": \\"object2.sourceProperty\\"
}"
`;

exports[`when merging object1 and object2 and two objects are merged with $merge in object2 then $targetProperty in object2 $merge.with should refer to the processed object2 $merge.source property 1`] = `
"{
	\\"targetProp\\": \\"object1.targetProp\\",
	\\"sharedProp\\": {
		\\"mergeSourceProp\\": \\"$merge.source.mergeSourceProp\\",
		\\"mergeSharedProp\\": \\"should be the value of mergeSharedProp\\",
		\\"mergeWithProp\\": \\"$merge.with.mergeWithProp\\"
	},
	\\"sourceProp\\": \\"object2.sourceProperty\\"
}"
`;

exports[`when merging object1 and object2 then $source in object2 should refer to the unprocessed object2 1`] = `
"{
	\\"targetProp\\": \\"object1.targetProp\\",
	\\"sharedProp\\": \\"should be the value of sharedProp\\",
	\\"sourceProp\\": \\"should be the value of sharedProp\\"
}"
`;

exports[`when merging object1 and object2 then $sourceProperty in object2 should refer to the unprocessed object2 property 1`] = `
"{
	\\"targetProp\\": \\"object1.targetProp\\",
	\\"sharedProp\\": \\"$sourceProperty\\"
}"
`;

exports[`when merging object1 and object2 then $target in object1 should refer to undefined 1`] = `"{}"`;

exports[`when merging object1 and object2 then $target in object2 should refer to the processed object1 1`] = `
"{
	\\"targetProp\\": \\"should be the value of sharedProp\\",
	\\"sharedProp\\": \\"should be the value of sharedProp\\"
}"
`;

exports[`when merging object1 and object2 then $targetProperty in object1 should refer to undefined 1`] = `"{}"`;

exports[`when merging object1 and object2 then $targetProperty in object2 should refer to the processed property in object1 1`] = `
"{
	\\"targetProp\\": \\"object1.targetProp\\",
	\\"sharedProp\\": \\"should be the value of sharedProp\\",
	\\"sourceProp\\": \\"object2.sourceProp\\"
}"
`;
